src/v/cloud_storage_clients/abs_client.cc

```diff
@@ -0,0 +1,600 @@
+    const boost::beast::string_view host{_ap().data(), _ap().length()};
+
+    http::client::request_header header{};
+    header.method(boost::beast::http::verb::put);
+    header.target(target);
+    header.insert(boost::beast::http::field::host, host);
+    header.insert(boost::beast::http::field::content_type, text_plain);
+    header.insert(
+      boost::beast::http::field::content_length,
+      std::to_string(payload_size_bytes));
+    header.insert("x-ms-blob-type", block_blob);
+
+    if (!tags.empty()) {
+        header.insert("x-ms-tags", tags.str());
+    }
+
+    auto error_code = _apply_credentials->add_auth(header);
+    if (error_code) {
+        return error_code;
+    }
+
+    return header;
+}
+
+result<http::client::request_header>
+abs_request_creator::make_get_blob_metadata_request(
+  bucket_name const& name, object_key const& key) {
+    // HEAD /{container-id}/{blob-id}?comp=metadata HTTP/1.1
+    // Host: {storage-account-id}.blob.core.windows.net
+    // x-ms-date:{req-datetime in RFC9110}
+    // x-ms-version:"2021-08-06"
+    // Authorization:{signature}
+    const auto target = fmt::format(
+      "/{}/{}?comp=metadata", name(), key().string());
+    const boost::beast::string_view host{_ap().data(), _ap().length()};
+
+    http::client::request_header header{};
+    header.method(boost::beast::http::verb::head);
+    header.target(target);
+    header.insert(boost::beast::http::field::host, host);
+
+    auto error_code = _apply_credentials->add_auth(header);
+    if (error_code) {
+        return error_code;
+    }
+
+    return header;
+}
+
+result<http::client::request_header>
+abs_request_creator::make_delete_blob_request(
+  bucket_name const& name, object_key const& key) {
+    // DELETE /{container-id}/{blob-id} HTTP/1.1
+    // Host: {storage-account-id}.blob.core.windows.net
+    // x-ms-date:{req-datetime in RFC9110}
+    // x-ms-version:"2021-08-06"
+    // Authorization:{signature}
+    const auto target = fmt::format("/{}/{}", name(), key().string());
+    const boost::beast::string_view host{_ap().data(), _ap().length()};
+
+    http::client::request_header header{};
+    header.method(boost::beast::http::verb::delete_);
+    header.target(target);
+    header.insert(boost::beast::http::field::host, host);
+    header.insert("x-ms-delete-snapshots", "include");
+
+    auto error_code = _apply_credentials->add_auth(header);
+    if (error_code) {
+        return error_code;
+    }
+
+    return header;
+}
+
+result<http::client::request_header>
+abs_request_creator::make_list_blobs_request(
+  const bucket_name& name,
+  std::optional<object_key> prefix,
+  [[maybe_unused]] std::optional<object_key> start_after,
+  std::optional<size_t> max_keys) {
+    // GET /{container-id}?restype=container&comp=list&prefix={prefix}...
+    // ...&max_results{max_keys}
+    // HTTP/1.1 Host: {storage-account-id}.blob.core.windows.net
+    // x-ms-date:{req-datetime in RFC9110}
+    // x-ms-version:"2021-08-06"
```

Comment:
That's right. `x-ms-version` defines the version of the API (i.e. parameters, limits, behaviour, etc.). It's currently hardcoded to `2021-08-06`. I chose this version as it's reasonably new, but not bleeding edge.

In terms of compat, my understanding is that these remain stable (aside from security issues perhaps), so I think we should be fine unless I'm misunderstanding your point.

Maybe it would be a good idea to make this property a tunable cluster config. It would make it easier to experiment with new versions and customer's won't have to upgrade their cluster when chancing the ABS API version.

Reply:
Makes sense, it'd be nice to have a comment saying what header fields are set by the auth layer, because there is a comment that lists all fields. I think that S3 client should have the same problem. It's not clear for the reader of the code what fields are set by what component. So, I guess it's a nit and can be ignored.

Reply:
Configuring `x-ms-verson` IMO is a good idea. We don't want to get stuck with something which will stop working eventually.

Reply:
I'll extend the comment to mention which header fields are added in the auth layer. Will make `x-ms-version` configurable too.

Reply:
It may be kind of dangerous if the user has modified this version, and we later add code that relies on a newer version -- we might modify the default API version, but users who have customized it will still be using older API version and our new code might not work.

Microsoft is still supporting 2014 version of their API, and our support window for Redpanda is only 1 year, so I don't think we need to worry too much about deprecated APIs in the near term.

I think it's sufficient to just pull the version in `signature_abs::sign_header` out somewhere a bit more prominent and add a comment explaining that this version might be incremented in future as needed, but is hardcoded to tie the API version we tested with to the API version we'll be using in the field.

Reply:
It's a potential footgun, but I still think it may be valuable to allow users to upgrade the API version without upgrading Redpanda. I was thinking the property should be tunable and we should specify in the docs what the current supported version is. I'm leaning towards adding this, because I don't see why users would ever want to change that without being prompted.

Reply:
I implemented this in a separate commit. See [here](https://github.com/redpanda-data/redpanda/pull/8086/commits/91b5f4c6fbb4a020751c03b528a76c268e1039c7). We can pull it out if we decide against it.

Reply:
Went for the hardcoded version in the end. John raised a good point against making it configurable. Consider the following scenario:
* Release Redpanda 23.1 with version X as the default
* Customer bumps config to version Y, at our request, for some issue.  We have confirmed that Redpanda 23.1 works with version Y, no problems yet.
* Later, we release Redpanda 24.1, which requires version Z.  Our new default is version Z.  The customer's config is still set to version Y, so when they install Redpanda 24.1, it can't use the new APIs that it expects to be able to use unless there is some manual intervention to reset the cluster's configuration so that it reverts to the default.

The problem in this case is that the API version won't be updated to version Z as the customer moved from cluster default in step 2.